lib/audit/test/audit.js

```diff
@@ -141,18 +150,134 @@ module.exports = {
+    if (!this.info.capability || !this.info.capability.candles) {
+      return
+    }
+
+    const markets = await au.get('/markets')
+    if (markets.length === 0) {
+      // If there are no markets, pass, since markets spec will fail
+      return
+    }
+
+    const market = markets[0]
+    const intervals = ['1d', '1h', '1m']
+
+    await Promise.all(
+      intervals.map(async interval => {
+        let candles
+
+        try {
+          candles = await au.get(`/candles?market=${encodeURIComponent(market.id)}&interval=${interval}`)
+        } catch (e) {
+          if (interval === '1d') {
+            au.assert(false, 'Expected to find candles endpoint for interval `1d`')
+          } else {
+            return // other candles endpoints are optional
+          }
+        }
+
+        const c = JSON.parse(JSON.stringify(candles)) // deep clone
+        let t = candles.map(c => c.timestamp)
+        let tSorted = c.map(c => c.timestamp).sort()
+
+        au.assert(
+          JSON.stringify(t) === JSON.stringify(tSorted),
+          `Expected ${interval} candles to be sorted by timestamp ascending`
+        )
+
+        if (interval === '1d') {
+          au.assert(candles.length >= 7, 'Expected at least 7 1d candles')
+          au.assert(
+            new Date(candles[candles.length - 1].timestamp).getTime() ===
+              moment()
+                .utc()
+                .startOf('day')
+                .toDate()
+                .getTime(),
+            `Expected last 1d candle to be current day UTC: ${candles[candles.length - 1].timestamp}`
+          )
+          au.assert(
+            new Date(candles[candles.length - 2].timestamp).getTime() ===
+              moment()
+                .utc()
+                .subtract(1, 'day')
+                .startOf('day')
+                .toDate()
+                .getTime(),
+            `Expected next to last 1d candle to be previous day UTC: ${candles[candles.length - 2].timestamp}`
+          )
+        }
+
+        if (interval === '1h') {
+          au.assert(candles.length >= 24, 'Expected at least 24 1h candles')
+          au.assert(
+            new Date(candles[candles.length - 1].timestamp).getTime() ===
+              moment()
+                .utc()
+                .startOf('hour')
+                .toDate()
+                .getTime(),
+            `Expected last 1h candle to be current hour UTC: ${candles[candles.length - 1].timestamp}`
+          )
+          au.assert(
+            new Date(candles[candles.length - 2].timestamp).getTime() ===
+              moment()
+                .utc()
+                .subtract(1, 'hour')
+                .startOf('hour')
+                .toDate()
+                .getTime(),
+            `Expected next to last 1h candle to be previous hour UTC: ${candles[candles.length - 2].timestamp}`
+          )
+        }
+
+        if (interval === '1m') {
+          au.assert(candles.length >= 60, 'Expected at least 60 1m candles')
```

Comment:
should we have the same minute-by-minute checking here as we do for hours and days?

Reply:
The minute-level checking was causing some audit failures. The minute candles are sometimes delayed by a minute (or so) so this was failing regularly for a few exchanges. We could add a fuzzy check that says the last 1m candle must be the current _or_ previous minute candle but I was hesitant to have the 1m "rule" deviate too much from the other intervals. Thoughts?

Reply:
It's OK if the "current" candle is not present. This is going to happen with all sizes simply due to slight difference in server clocks and lag creating the candle for the first time. I think it's most important to ensure the candles are ascending. For minute candles, lets say the most recent one should be within the past 60 minutes.